### PR TITLE
Add php-amqplib integration

### PR DESCRIPTION
**Pull request by PROFeNoM**
### Description

Add a php-amqplib integration (#1840).

**Methods Traced**

[AMQPChannel](https://github.com/php-amqplib/php-amqplib/blob/master/PhpAmqpLib/Channel/AMQPChannel.php)

- `PhpAmqpLib\Channel\AMQPChannel::basic_get`
- `PhpAmqpLib\Channel\AMQPChannel::basic_deliver`
- `PhpAmqpLib\Channel\AMQPChannel::basic_publish`
- `PhpAmqpLib\Channel\AMQPChannel::basic_cancel`
- `PhpAmqpLib\Channel\AMQPChannel::basic_cancel_ok`
- `PhpAmqpLib\Channel\AMQPChannel::basic_consume`
- `PhpAmqpLib\Channel\AMQPChannel::basic_consume_ok`
- `PhpAmqpLib\Channel\AMQPChannel::basic_ack`
- `PhpAmqpLib\Channel\AMQPChannel::basic_nack`
- `PhpAmqpLib\Channel\AMQPChannel::exchange_declare`
- `PhpAmqpLib\Channel\AMQPChannel::queue_declare`
- `PhpAmqpLib\Channel\AMQPChannel::queue_bind`

[AbstractConnection](https://github.com/php-amqplib/php-amqplib/blob/master/PhpAmqpLib/Connection/AbstractConnection.php)
- `PhpAmqpLib\Connection\AbstractConnection::connect`
- `PhpAmqpLib\Connection\AbstractConnection::reconnect`

### Readiness checklist
- [ ] (only for Members) Changelog has been added to the release document.
- [X] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: src/Integrations/Integrations/AMQP/AMQPIntegration.php [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+                $span->meta[Tag::MQ_OPERATION] = 'receive';
+                $span->meta[Tag::MQ_CONSUMER_ID] = $message->getConsumerTag();
+
+                $span->meta[Tag::RABBITMQ_EXCHANGE] = $exchangeDisplayName;
+                $span->meta[Tag::RABBITMQ_ROUTING_KEY] = $routingKeyDisplayName;
+
+                $integration->setOptionalMessageTags($span, $message);
+
+                // Try to extract propagated context values from headers
+                $integration->extract($message);
```
Comment (pierotibou) [inline src/Integrations/Integrations/AMQP/AMQPIntegration.php:75]: `extractContext` may be a better name

File: src/Integrations/Integrations/AMQP/AMQPIntegration.php [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+        );
+
+        trace_method(
+            "PhpAmqpLib\Channel\AMQPChannel",
+            "basic_publish",
+            [
+                'prehook' => function (SpanData $span, $args) use ($integration) {
+                    /** @var AMQPMessage $message */
+                    $message = $args[0];
+                    $integration->inject($message);
```
Comment (pierotibou) [inline src/Integrations/Integrations/AMQP/AMQPIntegration.php:86]: `injectContext`

File: src/Integrations/Integrations/AMQP/AMQPIntegration.php [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+                    $integration->setGenericTags($span, 'amqp.basic.publish', 'producer', $exception);
+
+                    /** @var AMQPMessage $message */
+                    $message = $args[0];
+                    /** @var string $exchange */
+                    $exchange = $args[1];
+                    /** @var string $routing_key */
+                    $routingKey = $args[2] ?? '';
+
+                    $exchangeDisplayName = empty($exchange) ? '<default>' : $exchange;
```
Comment (pierotibou) [inline src/Integrations/Integrations/AMQP/AMQPIntegration.php:98]: Can we factorize duplicated code?

File: src/Integrations/Integrations/AMQP/AMQPIntegration.php [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+            function (SpanData $span, $args) use ($integration) {
+                $integration->setGenericTags($span, 'amqp.basic.deliver', 'consumer');
+
+                /** @var AMQPMessage $message */
+                $message = $args[1];
+
+                $exchange = $message->getExchange();
+                $routingKey = $message->getRoutingKey();
+
+                $exchangeDisplayName = empty($exchange) ? '<default>' : $exchange;
```
[the remark this answers is not in the document]
Comment (PROFeNoM) [reply src/Integrations/Integrations/AMQP/AMQPIntegration.php:57]: This is indeed what I saw in the [Java](https://github.com/DataDog/dd-trace-java/blob/993e196dba2ca7d98881edffebdcd622e1984277/dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java#L105) and [Dotnet](https://github.com/DataDog/dd-trace-dotnet/blob/7c27a88f6047e96dd73b9c44e5d42a43c42d7f00/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/BasicPublishIntegration.cs#L57) instrumentations.

**Review by pierotibou — APPROVED** [review]

File: src/Integrations/Integrations/AMQP/AMQPIntegration.php [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+            }
+
+            // If present, set the sampling priority of the current span to x-datadog-sampling-priority
+            $priority = $headers['x-datadog-sampling-priority'] ?? null;
+            \DDTrace\set_priority_sampling($priority);
+
+            // If present, set the origin of the current span to x-datadog-origin
+            $origin = $headers['x-datadog-origin'] ?? null;
+
+            \DDTrace\set_distributed_tracing_context($traceId, $parentId, $origin);
```
Comment (bwoebi) [inline src/Integrations/Integrations/AMQP/AMQPIntegration.php:406]:
```suggestion
            DDTrace\consume_distributed_tracing_headers($headers);
```

File: src/Integrations/Integrations/AMQP/AMQPIntegration.php [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+        }
+        if ($message->has('message_id')) {
+            $span->meta[Tag::MQ_MESSAGE_ID] = $message->get('message_id');
+        }
+        if ($message->has('correlation_id')) {
+            $span->meta[Tag::MQ_CONVERSATION_ID] = $message->get('correlation_id');
+        }
+    }
+
+    public function injectContext(AMQPMessage $message)
```
Comment (bwoebi) [inline src/Integrations/Integrations/AMQP/AMQPIntegration.php:363]: Add a check for `DD_DISTRIBUTED_TRACING` config first.

File: src/Integrations/Integrations/AMQP/AMQPIntegration.php [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+        return empty($routingKey)
+            ? '<all>'
+            : (str_starts_with($routingKey, 'amq.gen-')
+                ? '<generated>'
+                : $routingKey);
+    }
+
+    public function setGenericTags(SpanData $span, string $name, string $spanKind, $exception = null)
+    {
+        $span->name = $name;
```
Comment (bwoebi) [inline src/Integrations/Integrations/AMQP/AMQPIntegration.php:334]:
```suggestion
    public function setGenericTags(SpanData $span, string $name, string $spanKind, string $resourceDetail = null, $exception = null)
    {
        $span->name = "amqp.$name";
        $span->resource = "$name" . ($resourceDetail === null ? "" : " $resourceDetail");
```

Wouldn't that reduce redundancy?

File: src/Integrations/Integrations/AMQP/AMQPIntegration.php [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+                    /** @var AMQPMessage $message */
+                    $exchange = $message->getExchange();
+                    $routingKey = $message->getRoutingKey();
+
+                    $exchangeDisplayName = empty($exchange) ? '<default>' : $exchange;
+                    $routingKeyDisplayName = empty($routingKey)
+                        ? '<all>'
+                        : (str_starts_with($routingKey, 'amq.gen-')
+                            ? '<generated>'
+                            : $routingKey);
```
Comment (bwoebi) [inline src/Integrations/Integrations/AMQP/AMQPIntegration.php:293]:
```suggestion
                        : $this->formatQueueName($routingKey);
```

File: tests/Integrations/AMQP/AMQPTest.php [hunk 9]
```diff
@@ -0,0 +1,837 @@
...
+                Tag::MQ_PROTOCOL                => 'AMQP',
+                Tag::MQ_PROTOCOL_VERSION        => AMQPChannel::getProtocolVersion()
+            ]),
+        ]);
+    }
+
+    public function testDistributedTracing()
+    {
+        // Note: This test is extremely flaky, locally at least. It will eventually pass with some tries...
+        // Reason: We may parse the traces from dumped data BEFORE the traces are flushed.
```
Comment (bwoebi) [inline tests/Integrations/AMQP/AMQPTest.php:798]: Can you elaborate? Also `\DDTrace\Tests\Common\TracerTestTrait::resetRequestDumper()` is a thing, if needed.

**Review by bwoebi — APPROVED** [review]

File: tests/Integrations/AMQP/AMQPTest.php (hunk 9 above)
Comment (bwoebi) [reply tests/Integrations/AMQP/AMQPTest.php:798]: For me it was never flaky, so, let's ship it.

**Pull request merged**